File: sdks/go/loreweave_mcp/confirm_token.go
package loreweave_mcp

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Confirm-token spine (Tier-S/W). A stateless HMAC over the claims, keyed by a
// per-service secret with a domain separator so the token can never be confused
// with a real JWT. Forging one requires the secret — i.e. a full service
// compromise, out of scope.
//
// The token binds INTENT (descriptor + payload) + IDENTITY (user) + RESOURCE +
// EXPIRY. The propose tool mints it (no write); the per-domain
// /v1/<domain>/actions/confirm route verifies it and is the ONLY write path
// (INV-9). The descriptor is the confused-deputy guard: confirm MUST check that
// the token's descriptor matches the action being confirmed, so a token minted
// for "book.delete" can never confirm "book.publish".
//
// Single-use, if required, is enforced at confirm time by the service (e.g.
// recording a hash of the token, or the (user,resource,descriptor,exp) tuple) —
// it is not in the stateless token itself.
//
// COMPOSE-A reconciliation note: this scheme is the canonical confirm spine
// shared with the Python kit (sdks/python/loreweave_mcp) at the API + claim
// level — same signature `(secret, user, resource, descriptor, payload, ttl)`,
// same claim fields, same domain separator. Tokens are minted AND verified
// inside one service (never across languages), so byte-level wire interop is
// deliberately not a contract; the alignment is conceptual.
const confirmTokenDomain = "lw-action-confirm:v1|"

var (
	// ErrConfirmTokenInvalid — bad signature, malformed, or empty secret.
	ErrConfirmTokenInvalid = errors.New("confirm token is invalid")
	// ErrConfirmTokenExpired — signature valid but past expiry (distinct so the
	// UI can say "re-propose" rather than "tampered").
	ErrConfirmTokenExpired = errors.New("confirm token has expired")
)

// ConfirmClaims is the verified payload of a confirm token. Payload is the opaque
// action-spec captured at propose time (resolved ids, validated values); confirm
// trusts it because it is inside the HMAC, but the service SHOULD still re-check
// user/resource against the live request and re-validate the payload against
// current state before executing.
type ConfirmClaims struct {
	UserID     uuid.UUID       `json:"u"`           // the proposing user (re-checked at confirm)
	ResourceID uuid.UUID       `json:"r"`           // the resource the action targets
	Descriptor string          `json:"d"`           // intent (e.g. "book.publish") — confirm MUST match this
	Payload    json.RawMessage `json:"p,omitempty"` // opaque action-spec
	Exp        int64           `json:"exp"`         // unix seconds
}

// MintConfirmToken signs a confirm token binding userID + resourceID + descriptor
// + payload, stamped with now+ttl. The secret is supplied by the caller (loaded
// from the service's own env — never a literal here); an empty secret fails
// closed with ErrConfirmTokenInvalid (the service must be configured to mint).
func MintConfirmToken(secret string, userID, resourceID uuid.UUID, descriptor string, payload any, ttl time.Duration) (string, error) {
	if secret == "" {
		return "", ErrConfirmTokenInvalid
	}
	var raw json.RawMessage
	if payload != nil {
		b, err := json.Marshal(payload)
		if err != nil {
			return "", ErrConfirmTokenInvalid
		}
		raw = b
	}
	claims := ConfirmClaims{
		UserID:     userID,
		ResourceID: resourceID,
		Descriptor: descriptor,
		Payload:    raw,
		Exp:        time.Now().Add(ttl).Unix(),
	}
	body, err := json.Marshal(claims)
	if err != nil {
		return "", ErrConfirmTokenInvalid
	}
	payloadB64 := base64.RawURLEncoding.EncodeToString(body)
	sig := confirmTokenSign(secret, payloadB64)
	return payloadB64 + "." + base64.RawURLEncoding.EncodeToString(sig), nil
}

// VerifyConfirmToken checks the HMAC signature (constant-time) then expiry, using
// the caller-supplied secret. A bad signature / format / empty secret →
// ErrConfirmTokenInvalid; a signature-valid but stale token →
// ErrConfirmTokenExpired. On success returns the bound claims; the caller MUST
// re-check user/resource against the live request, confirm the descriptor matches
// the action, and (if one-shot) record single-use.
func VerifyConfirmToken(secret, tok string) (ConfirmClaims, error) {
	var zero ConfirmClaims
	if secret == "" {
		return zero, ErrConfirmTokenInvalid
	}
	parts := strings.Split(tok, ".")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return zero, ErrConfirmTokenInvalid
	}
	sig, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return zero, ErrConfirmTokenInvalid
	}
	expected := confirmTokenSign(secret, parts[0])
	if subtle.ConstantTimeCompare(sig, expected) != 1 {
		return zero, ErrConfirmTokenInvalid
	}
	body, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return zero, ErrConfirmTokenInvalid
	}
	var claims ConfirmClaims
	if err := json.Unmarshal(body, &claims); err != nil {
		return zero, ErrConfirmTokenInvalid
	}
	if time.Now().Unix() >= claims.Exp {
		return zero, ErrConfirmTokenExpired
	}
	return claims, nil
}

func confirmTokenSign(secret, payloadB64 string) []byte {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(confirmTokenDomain))
	mac.Write([]byte(payloadB64))
	return mac.Sum(nil)
}

File: sdks/python/loreweave_mcp/confirm_token.py
"""Confirm-token spine (INV-9) — PORTED from the Go/glossary scheme in
``services/glossary-service/internal/api/action_confirm_token.go`` so a token
minted by the Go kit verifies in the Python kit and vice-versa.

WIRE FORMAT (must stay byte-identical to the Go kit — S-KIT-GO):

    token := base64url(payload_json) + "." + base64url(hmac_sha256(secret, domain || payloadB64))

  - base64url = URL-safe base64, NO padding (Go ``base64.RawURLEncoding``).
  - signature = HMAC-SHA256 keyed by the service secret, over the bytes
    ``DOMAIN || payloadB64`` where ``DOMAIN = "lw-action-confirm:v1|"`` is a
    domain separator written BEFORE the payload (Go writes the domain then the
    payloadB64 into the same HMAC — `actionTokenSign`). The separator guarantees a
    confirm token can never be confused with a real JWT.
  - constant-time signature compare on verify; signature/format/unknown → INVALID;
    valid-but-stale → EXPIRED (distinct, so the UI says "re-propose" not "denied").

CLAIM JSON SHAPE (compact keys — must match Go ``actionClaims`` tags exactly):

    { "u":   "<user_id uuid>",        // the proposing user (identity binding)
      "r":   "<resource_id uuid>",    // book/model/record the action targets
      "d":   "<descriptor>",          // e.g. "book.publish_batch", "translation.start_job"
      "p":   <opaque payload json>,   // the resolved action-spec captured at propose time
      "exp": <unix seconds> }         // absolute expiry

The token binds intent (descriptor) + identity (u) + resource (r) + payload (p) +
expiry. Single-use is NOT in the token (it is stateless) — a consuming service that
needs one-shot semantics records the (u, r, d, exp) in its own consumed-token ledger
at confirm time. The kit provides only the stateless bind+verify.

NOTE for Go/Py alignment: the Go glossary file uses ``jti`` + ``auth`` + ``b``
(book) keys for its richer scheme. The FROZEN C-KIT shape here is the SMALLER,
domain-agnostic ``{u, r, d, p, exp}`` (resource-generic, not book-specific) so the
SAME claim shape serves book / settings / translation. S-KIT-GO must emit/verify
THIS shape (see the report note for reconciliation).
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass
from typing import Any
from uuid import UUID

__all__ = [
    "ConfirmClaims",
    "ConfirmTokenError",
    "ConfirmTokenInvalid",
    "ConfirmTokenExpired",
    "mint_confirm_token",
    "verify_confirm_token",
]

# Domain separator — MUST match the Go kit byte-for-byte. (Distinct from the
# glossary file's legacy "gloss-action-confirm:v1|" because the C-KIT spine is
# the platform-wide, domain-agnostic shape; S-KIT-GO uses this same constant.)
_DOMAIN = b"lw-action-confirm:v1|"

DEFAULT_TTL_S = 600  # 10 minutes — human has time to read + confirm (mirrors Go)


class ConfirmTokenError(Exception):
    """Base for confirm-token failures."""


class ConfirmTokenInvalid(ConfirmTokenError):
    """Signature/format mismatch, or required claim missing/malformed."""


class ConfirmTokenExpired(ConfirmTokenError):
    """A structurally-valid, correctly-signed token whose ``exp`` has passed."""


@dataclass(frozen=True)
class ConfirmClaims:
    user_id: UUID
    resource_id: UUID
    descriptor: str
    payload: Any
    exp: int  # unix seconds


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _b64url_decode(s: str) -> bytes:
    # Re-pad for the stdlib decoder (Go RawURLEncoding emits no padding).
    pad = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


def _sign(secret: str, payload_b64: str) -> bytes:
    mac = hmac.new(secret.encode("utf-8"), digestmod=hashlib.sha256)
    mac.update(_DOMAIN)
    mac.update(payload_b64.encode("ascii"))
    return mac.digest()


def mint_confirm_token(
    secret: str,
    user_id: UUID,
    resource_id: UUID,
    descriptor: str,
    payload: Any,
    ttl: int = DEFAULT_TTL_S,
    *,
    now: float | None = None,
) -> str:
    """Mint a stateless HMAC confirm token binding (user, resource, descriptor,
    payload) with an absolute expiry ``now + ttl``.

    An empty secret or descriptor is a misconfiguration → raises
    ``ConfirmTokenInvalid`` (caller treats "cannot mint" as fail-closed, never
    issuing an unsigned token). ``now`` is injectable for tests.
    """
    if not secret:
        raise ConfirmTokenInvalid("cannot mint: empty secret")
    if not descriptor or not descriptor.strip():
        raise ConfirmTokenInvalid("cannot mint: empty descriptor")
    issued = time.time() if now is None else now
    claims = {
        "u": str(user_id),
        "r": str(resource_id),
        "d": descriptor,
        "p": payload,
        "exp": int(issued) + int(ttl),
    }
    # Sort keys for a deterministic encoding (helps cross-impl reproducibility of
    # logs/tests; the signature does not require it but determinism is cheap).
    payload_json = json.dumps(claims, separators=(",", ":"), sort_keys=True)
    payload_b64 = _b64url_encode(payload_json.encode("utf-8"))
    sig = _sign(secret, payload_b64)
    return payload_b64 + "." + _b64url_encode(sig)


def verify_confirm_token(
    secret: str, token: str, *, now: float | None = None
) -> ConfirmClaims:
    """Verify a confirm token: constant-time signature check, required-claim
    validation, then expiry.

    - bad secret / format / signature / missing-or-malformed claim →
      ``ConfirmTokenInvalid``.
    - correctly-signed but stale → ``ConfirmTokenExpired`` (distinct so the UI can
      say "re-propose").
    """
    if not secret:
        raise ConfirmTokenInvalid("empty secret")
    parts = token.split(".")
    if len(parts) != 2 or not parts[0] or not parts[1]:
        raise ConfirmTokenInvalid("malformed token")

    payload_b64, sig_b64 = parts
    try:
        sig = _b64url_decode(sig_b64)
    except (ValueError, base64.binascii.Error):
        raise ConfirmTokenInvalid("malformed signature")

    expected = _sign(secret, payload_b64)
    if not hmac.compare_digest(sig, expected):
        raise ConfirmTokenInvalid("signature mismatch")

    try:
        payload_raw = _b64url_decode(payload_b64)
        claims = json.loads(payload_raw)
    except (ValueError, base64.binascii.Error):
        raise ConfirmTokenInvalid("malformed payload")
    if not isinstance(claims, dict):
        raise ConfirmTokenInvalid("malformed payload")

    try:
        user_id = UUID(str(claims["u"]))
        resource_id = UUID(str(claims["r"]))
        descriptor = str(claims["d"])
        exp = int(claims["exp"])
    except (KeyError, ValueError, TypeError):
        raise ConfirmTokenInvalid("missing or malformed claim")
    if not descriptor.strip():
        raise ConfirmTokenInvalid("empty descriptor")

    current = time.time() if now is None else now
    if int(current) >= exp:
        raise ConfirmTokenExpired("token has expired")

    return ConfirmClaims(
        user_id=user_id,
        resource_id=resource_id,
        descriptor=descriptor,
        payload=claims.get("p"),
        exp=exp,
    )
